KEP:  Draft For Repository (targeted for KEP-15) (#654)
kudobuilder/kudo: keps/draft-repo.md

      * [Proposal](#proposal)
         * [Create Operator Tarball](#create-operator-tarball)
         * [Create Repository Index](#create-repository-index)
         * [Create Local Respository](#create-local-respository)
         * [Update a Repository](#update-a-repository)
         * [CLI Repository Help](#cli-repository-help)
         * [User Stories](#user-stories)
         * [Risks and Mitigations](#risks-and-mitigations)
      * [Graduation Criteria](#graduation-criteria)

## Summary

There is a number of ways that an operator can be installed via KUDO.  The operator developer can install via the local file system (with a operator folder or tarball).  It is also possible to install via url to a tarball.  While [KEP-0010](keps/0010-package-manager.md) defines the packaging structure, this kep defines:

* The structure of a repository
* Defining authoritative repositories: incubator & graduated
* Working with multiple repositories
* Packaging and Repository management tooling
* Configuration of repositories for the CLI
* CLI Tooling for search, packaging, and pushing to/from repository
* Define the rules and means to promote incubated operators to the graduated repository.

> By providing a repository (or set of repositories), KUDO will provide a mechanism for administrators to easy discover and install KUDO operators.  [Helm has provided some prior art](https://github.com/helm/helm/blob/master/docs/chart_repository.md) which we should align closely to from a mental model stand point.

Reviewer: So.. while we're taking a note from Helm's book.. Helm is actually moving away from the custom chart repository model: https://blog.bacongobbler.com/post/2019-01-25-distributing-with-distribution/index.html

So they outline a number of problems with the custom chart repository model:

> While there are several benefits to the Chart Repository API meeting the most basic storage requirements, a few drawbacks have started to show:
> * Chart Repositories have a very hard time abstracting most of the security implementations required in a production environment. Having a standard API for authentication and authorization is very important in production scenarios
> * Helm’s Chart provenance tools used for signing and verifying the integrity and origin of a chart is an after-thought of the chart publishing process
> * In multi-tenant scenarios, the same chart can be uploaded by another tenant, costing twice the storage cost to store the same content. Smarter chart repositories can be designed to handle this, but it’s not a part of the spec
> * Using a single index file for search, metadata information and fetching charts has made it difficult or clunky to design around in secure multi-tenant implementations.

Compared to using OCI-compatible registries for a backend:

> Let’s talk about what the Distribution project brings to the table:
> * A stable, battle-hardened product with many cloud vendors providing their own flavour of Distribution
> * A rich set of libraries for interacting with Distribution
> * A pluggable storage backend, including support for local filesystem, Azure, GCS, Aliyun OSS, AWS S3, and Openstack Swift out of the box
> * A standard API for authentication/authorization
> * Content stored in Distribution is content-addressable via a sha256 checksum. In other words, if two files are uploaded with identical content, only one copy of the content is stored.

So it seems like quite a bit of the work is already done for us in the OCI image format and repositories. Maybe we should explore that more?

Author: totally

Reviewer 2: Oh niiiice, why haven't I seen this before!! :) Couple of thoughts:
- I think our current model is fine up until the point where we want to support people running their own registries (eg. airgaps? other use cases?) then we basically run to the same "problems" as they note there
- the downside of using repository model is that you actually have to run the repository 'application' somewhere and manage it while right now we just use hosted GCS and don't need to maintain official repo app that has to be up all the time

all and all I think we should investigate both options and probably write pros and cons to this list. Maybe we can stick with current model for now and move to that one later? (it worked for helm for quite some time) But maybe that does not make sense if we would invest a lot of time in the current model that will just get thrown away.

Also not sure what that means for the V2 deliverable next friday :)

## Motivation

> We need a way for the operator author to update a repository with his/her operator.  We also need a way for the operator to promote an operator as being production ready. Finally we need a KUDO user to point to different repositories.  

Reviewer: ```suggestion
We need a way for the operator author to update a repository with their operator.  We also need a way for the operator to promote an operator as being production ready. Finally we need a KUDO user to point to different repositories.  
```

### Goals

- Automate the creation of a operator tarball
- Automate the creation or updating of the repository index with a new operator tarball
- Define rules for incubator and graduated along with graduation tooling.
- Define how an admin can create their own repository
- Install operator from incubator instead of graduated repository

### Non-Goals

- Manage repos of non-KUDO applications (e.g. Helm Charts)
- External formats defined by [KEP0013](keps/0013-external-specs.md)
- Repository cacheing

## Proposal

### Create Operator Tarball

> An operator package / bundle as defined in [KEP0010](keps/0010-package-manager.md).  The storage package at the repository is a tarball. We need a way to create this tarball in a uniformed way. KUDO shall a way to generate a tarball based on the standard file system layout.  As an example: `kubectl kudo package docs/examples/zookeeper/`.  This will create a `zookeeper-3.4.10.tgz` based on parsing the `operator.yaml` for version details. It would be best if this included linting to ensure that the static structure of the operator is correct. At this point, it will guarantee that operator folder has `operator.yaml` and `params.yaml` along with a templates directory.  It will create the following operator tarball `{operator.yaml:name}-{operator.yaml:version}.tgz`.

Reviewer: We say "package / bundle" here, but didn't we agree on a term?

### Create Repository Index

> KUDO needs the ability to create an index file for the repository. Something like `kubectl kudo repo index new-repo --url https://kudo-repo.storage.googleapis.com`. In this example, the `kudo repo index` is the command. The `new-repo` is a folder containing operator tarballs. The index file defined in [KEP0010](keps/0010-package-manager.md) will be created using the url provided for links in the file.

Reviewer: Could the URLs be relative instead of absolute?

Author: seems like an implementation detail that I'm not entire sure about at this point.  This kep is to capture the need to create a repo index.   if we can find ways to make life easier 👍.   The helm docs used as inspiration does not.

Steps for creating an index file for a new operator looks like:
```

> kubectl kudo package docs/examples/zookeeper/

Reviewer: Maybe this command could specify an output directory?

```
kubectl kudo package --directory new-repo docs/examples/zookeeper/
kubectl kudo repo index new-repo --url https://kudo-repo.storage.googleapis.com
```

Author: yeah... I was thinking the same.   not all flags were thought of in the draft of the kep... that's a good suggestion that I will add.

Author: although in this case... the implementation actually had the feature of adding `destination` so we are definitely on the same page.   see #677

mkdir new-repo
kubectl kudo package docs/examples/zookeeper/ --destination=new-rep
kubectl kudo repo index new-repo --url https://kudo-repo.storage.googleapis.com
```

> ### Create Local Respository

Reviewer 2: this is for me out of scope for this KEP, I would remove it. It does not even address any of the goals. I think we need to put a more thought into running local/custom repositories, I would almost see that as a separate KEP

Reviewer 2: I would love our keps to be very lean and something we commit to in one release iteration and I don't see this as an essential feature, just nice to have. Does not really help with Day 2 or our core business so I would postpone it and remove it but if you feel like it's very high on our roadmap 🤷‍♀ just keep in mind every feature mean implementing it, testing it and documenting it

For convenience it should be possible to have KUDO start a repository service with a provided path.  `kubectl kudo serve --repo-path ./repo` would create a local http service for the repo in `./repo`

### Update a Repository

> Inspired by Helm, KUDO needs a way to take a repository index and merge it with another index for the purposes of adding operators to a repository.  Example: `kubectl kudo  repo --merge`.

Reviewer 2: why is --merge a flag and not a command? 🤔 also what are we merging here? It expects to be run locally on filesystem with that indexfile and repo?

### CLI Repository Help

> `kubectl kudo  repo index --help`

Reviewer 2: shouldn't this be `kudo repo --help`?

Author: we support `kudo repo --help` and `kudo repo index --help`

### User Stories

- Allow pushing of new KUDO operator to the incubator repository
- Allow running of a CNAB bundle as a KUDO Operator
- Allow the running of an Operator as a KUDO Operator

### Risks and Mitigations

## Graduation Criteria

> We need to define the rules for promoting incubator operators to the graduated repository.

Reviewer: I think at least one criteria would be some level of test coverage and passing in the Kubernetes environments defined in KEP-0004.
